ddtrace/internal/module.py

```diff
@@ -0,0 +1,482 @@
+        # type: (object) -> bool
+        return self._modules.__contains__(name)
+
+    def __len__(self):
+        # type: () -> int
+        return self._modules.__len__()
+
+    def __iter__(self):
+        # type: () -> Iterator
+        return self._modules.__iter__()
+
+    def find_module(self, fullname, path=None):
+        # type: (str, Optional[str]) -> Union[ModuleWatchdog, _ImportHookChainedLoader, None]
+        if fullname in self._finding:
+            return None
+
+        self._finding.add(fullname)
+
+        try:
+            if PY2:
+                __import__(fullname)
+                return _ImportHookLoader(self.after_import)
+
+            loader = getattr(find_spec(fullname), "loader", None)
+            if loader and not isinstance(loader, _ImportHookChainedLoader):
+                return _ImportHookChainedLoader(loader, self.after_import)
+
+        finally:
+            self._finding.remove(fullname)
+
+        return None
+
+    def find_spec(self, fullname, path=None, target=None):
+        # type: (str, Optional[str], Optional[ModuleType]) -> Optional[ModuleSpec]
+        if fullname in self._finding:
+            return None
+
+        self._finding.add(fullname)
+
+        try:
+            spec = find_spec(fullname)
+            if spec is None:
+                return None
+
+            loader = getattr(spec, "loader", None)
+
+            if loader and not isinstance(loader, _ImportHookChainedLoader):
+                spec.loader = _ImportHookChainedLoader(loader, self.after_import)
+
+            return spec
+
+        finally:
+            self._finding.remove(fullname)
+
+    @classmethod
+    def register_origin_hook(cls, origin, hook, arg):
+        # type: (str, HookType, Any) -> None
+        """Register a hook to be called when the module with the given origin is
+        imported.
+
+        The hook will be called with the module object and the given argument.
+        """
+        cls._check_installed()
+
+        # DEV: Under the hypothesis that this is only ever called by the probe
+        # poller thread, there are no further actions to take. Should this ever
+        # change, then thread-safety might become a concern.
+        path = _resolve(origin)
+        if path is None:
+            raise ValueError("Cannot resolve module origin %s" % origin)
+
+        log.debug("Registering hook '%r' on path '%s'", hook, path)
+        instance = cast(ModuleWatchdog, cls._instance)
+        instance._hook_map[path].append((hook, arg))
+        try:
+            module = instance._origin_map[path]
+        except KeyError:
+            # The module is not loaded yet. Nothing more we can do.
+            return
+
+        # The module was already imported so we invoke the hook straight-away
+        log.debug("Calling hook '%r' on already imported module '%s'", hook, module.__name__)
+        hook(module, arg)
+
+    @classmethod
+    def unregister_origin_hook(cls, origin, arg):
```

Comment:
```suggestion
    def deregister_origin_hook(cls, origin, arg):
```

just a nit/convention we follow: https://github.com/DataDog/dd-trace-py/blob/892b292ba7054b2b529bda6738d0a39de5f7b9a4/ddtrace/tracer.py#L251

Reply:
🤔  this doesn't seem to be enforced at the moment though
https://github.com/DataDog/dd-trace-py/blob/892b292ba7054b2b529bda6738d0a39de5f7b9a4/ddtrace/internal/atexit.py#L47
https://github.com/DataDog/dd-trace-py/blob/892b292ba7054b2b529bda6738d0a39de5f7b9a4/ddtrace/internal/forksafe.py#L50